### PR TITLE
R4R: refactor auth and bank

### PR DESCRIPTION
**Pull request by HaoyangLiu**

### CODE REVIEW

File: modules/auth/keeper.go [hunk 1]
```diff
@@ -1,12 +1,23 @@
...
+
 	codec "github.com/irisnet/irishub/codec"
 	sdk "github.com/irisnet/irishub/types"
 	"github.com/tendermint/tendermint/crypto"
 )
 
-var globalAccountNumberKey = []byte("globalAccountNumber")
+var (
+	// Prefix for account-by-address store
+	addressStoreKeyPrefix = []byte{0x01}
```
Comment (chengwenxi) [inline modules/auth/keeper.go]: to []byte("account:")

File: modules/auth/keeper.go [hunk 2]
```diff
@@ -160,6 +167,128 @@ func (am AccountKeeper) GetNextAccountNumber(ctx sdk.Context) uint64 {
 	return accNumber
 }
 
+func (am AccountKeeper) GetBurnToken(ctx sdk.Context) sdk.Coins {
```
Comment (wukongcheng) [inline modules/auth/keeper.go]: GetBurnedToken

File: modules/auth/keeper.go [hunk 3]
```diff
@@ -160,6 +167,128 @@ func (am AccountKeeper) GetNextAccountNumber(ctx sdk.Context) uint64 {
...
+	store := ctx.KVStore(am.key)
+	bz := store.Get(burnTokenKey)
+	if bz == nil {
+		burnToken = nil
+	} else {
+		err := am.cdc.UnmarshalBinaryLengthPrefixed(bz, &burnToken)
+		if err != nil {
+			panic(err)
+		}
+	}
```
Comment (wukongcheng) [inline modules/auth/keeper.go]: call the above function `GetBurnToken` instead of the duplicated codes

File: modules/auth/keeper.go [hunk 4]
```diff
@@ -160,6 +167,128 @@ func (am AccountKeeper) GetNextAccountNumber(ctx sdk.Context) uint64 {
...
+	// increase totalLoosenToken
+	burnToken = burnToken.Plus(coins)
+	if !burnToken.IsNotNegative() {
+		panic(fmt.Errorf("burn token is negative"))
+	}
+	// write back to db
+	bzNew, err := am.cdc.MarshalBinaryBare(burnToken)
```
Comment (wukongcheng) [inline modules/auth/keeper.go]: MarshalBinaryLengthPrefixed ?

File: modules/auth/keeper.go [hunk 5]
```diff
@@ -160,6 +167,128 @@ func (am AccountKeeper) GetNextAccountNumber(ctx sdk.Context) uint64 {
...
+	store := ctx.KVStore(am.key)
+	bz := store.Get(totalLoosenTokenKey)
+	if bz == nil {
+		totalLoosenToken = nil
+	} else {
+		err := am.cdc.UnmarshalBinaryLengthPrefixed(bz, &totalLoosenToken)
+		if err != nil {
+			panic(err)
+		}
+	}
```
Comment (wukongcheng) [inline modules/auth/keeper.go]: call the above function GetTotalLoosenToken instead of the duplicated codes

File: modules/auth/keeper.go [hunk 6]
```diff
@@ -160,6 +167,128 @@ func (am AccountKeeper) GetNextAccountNumber(ctx sdk.Context) uint64 {
...
```
Comment (wukongcheng) [inline modules/auth/keeper.go]: call the above function GetTotalLoosenToken instead of the duplicated codes

File: modules/auth/keeper.go [hunk 7]
```diff
@@ -160,6 +167,100 @@ func (am AccountKeeper) GetNextAccountNumber(ctx sdk.Context) uint64 {
...
 
+func (am AccountKeeper) GetBurnedToken(ctx sdk.Context) sdk.Coins {
+	// read from db
+	var burnToken sdk.Coins
+	store := ctx.KVStore(am.key)
+	bz := store.Get(burnTokenKey)
+	if bz == nil {
+		burnToken = nil
+	} else {
+		err := am.cdc.UnmarshalBinaryLengthPrefixed(bz, &burnToken)
```
Comment (chengwenxi) [inline modules/auth/keeper.go]: Use MustUnmarshalBinaryLengthPrefixed

**Pull request merged**